### PR TITLE
Batch size bucketing

### PR DESCRIPTION
**Pull request by kdamaszk**
# What does this PR do?

This PR changes the server logic towards static shapes in the context of batch size.
Thanks to this, we are able to create batches of a fixed size, which saves memory (fewer graphs in memory), and also makes the process of merging and deleting requests more developed in terms of further optimization. This is a key feature from the point of view of continuous batching.

### CODE REVIEW

**Comment by kdamaszk** [comment]
@regisss @libinta Please take a look on this change. It is a crucial feature from continuous batching perspective.

File: server/text_generation_server/models/causal_lm.py [hunk 1]
```diff
@@ -719,18 +531,19 @@ def forward(
...
 
         if self.is_optimized_for_gaudi:
-            token_idx = torch.tensor(batch.attention_mask.shape[-1] - batch.padding_right_offset).to(self.device)
+            token_idx = torch.tensor(batch.attention_mask.shape[-1] - batch.right_padding).to(self.device)
             attention_mask = batch.attention_mask
-
         else:
             token_idx = None
             # slice the attention mask to the correct shape
+            # TODO fix me!
```
Comment (regisss) [inline server/text_generation_server/models/causal_lm.py:546]: Is it something to fix in this PR?

**Review by regisss — APPROVED** [review]
Thanks!

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (kdamaszk) [reply server/text_generation_server/models/causal_lm.py:546]: No. This has a low priority, as we are aiming in is_optimized_for_gaudi case.
Summing up, I know there are a few TODOs in this code, but these are mostly related with future work.

**Pull request merged**